### ISSUE TITLE
Pydantic outer expected type propagates through sorted() into lambda inference

### ISSUE THREAD

**Issue by sunfkny** [astral-sh/ty#4562]
### Summary

```
from pydantic import BaseModel


class Person(BaseModel):
    uid: str
    name: str


class SearchResponse(BaseModel):
    people: list[Person]


people_raw: list[Person] = []
demo = SearchResponse(people=sorted(people_raw, key=lambda p: p.uid))
```

With the latest ty 0.0.82:

```
> uvx ty@0.0.82 check demo.py
error[unresolved-attribute]: Attribute `uid` is not defined on `Mapping[str, Any]` in union `Person | Mapping[str, Any]`
  --> demo.py:14:63
   |
14 | demo = SearchResponse(people=sorted(people_raw, key=lambda p: p.uid))
   |                                                               ^^^^^

Found 1 diagnostic
```

## Expected behavior

No diagnostic should be emitted.

`people_raw` is explicitly typed as `list[Person]`, so the element type of `sorted()` should be inferred as `Person`. Therefore, the parameter of the key lambda should also be inferred as `Person`.

## Actual behavior

The lambda parameter appears to be inferred as:

`Person | Mapping[str, Any]`

instead of `Person`, resulting in:

```
Attribute `uid` is not defined on `Mapping[str, Any]` in union `Person | Mapping[str, Any]`
```

The `Mapping[str, Any]` part appears to originate from the input type accepted by Pydantic for the `Person` model.

## Outer expected type appears to propagate through `sorted()`

The error disappears if the `sorted()` expression is first assigned to a local variable:

```
people = sorted(people_raw, key=lambda p: p.uid)
demo = SearchResponse(people=people)
```

This suggests that the expected type from `SearchResponse(people=...)` propagates through `sorted()` and affects the inferred parameter type of its key lambda.

Since `people_raw` is explicitly typed as `list[Person]`, the iterable argument already provides a concrete element type for `sorted()`. I would therefore expect the lambda parameter to remain `Person` rather than being widened by the outer Pydantic input type.

## Additional context

I can reproduce this with both ty `0.0.63` and the latest ty `0.0.82`.

ty `0.0.62` does not report this diagnostic, but it predates the relevant lambda contextual inference support.

### Version

ty 0.0.82 (6999edb88 2026-09-17)

**Comment by sharkdp** [comment]
Thank you. We're tracking this in https://github.com/astral-sh/ty/issues/4016
